### PR TITLE
SSH parsing blog post

### PR DESCRIPTION
**Pull request by BenjaminGrayNp1**
Why not

### CODE REVIEW

File: content/blog/2023-08-03-parsing-ssh-config.markdown [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+config correctly? If it was trying to connect directly from the host, it would
+make sense to fail a DNS lookup.
+
+
+## Investigating
+
+Enough theorising, time to debug the extension as it tries to connect.
+
+From the error above, the string `"Error resolving authority"` looks like
+something I can search for. This takes me to the `catch` case for a large
```
Comment (ruscur) [inline content/blog/2023-08-03-parsing-ssh-config.markdown]: Link to the source here?

File: content/blog/2023-08-03-parsing-ssh-config.markdown [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+second go around. This will pass, and, since nothing has set `ProxyJump` yet, we
+would gain the proxy.
+
+You may think, yes, but we don't have a `Match final` in that example. But if
+you thought that, then you forgot about the system config.
+
+The system config is effectively appended to the user config, to allow any
+system wide settings. Most of the time this isn't an issue because of the
+first-come-first-served rule with config matches (rule 2). But if the system
+config includes a `Match final`, it will trigger the entire config to be
```
Comment (ruscur) [inline content/blog/2023-08-03-parsing-ssh-config.markdown:158]: I would mention that on Fedora that's in /etc/ssh/ssh_config.d, so this is a real thing that actually happens

File: content/blog/2023-08-03-parsing-ssh-config.markdown [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+Enter `ssh -G`. The `-G` flag asks SSH to dump the complete resolved config,
+without actually opening the connection (it may execute arbitrary code while
+resolving the config however!). So to fix the extension once and for all, we
+could swap the manual parser to just invoking `ssh -G example`, and parsing the
+output as the final config. No `Host` or `Match` or `HostName` or `Match final`
+quirks to worry about.
+
+Sure enough, if we replace the config backend with this 'native' resolver, we
+can connect to all the machines with no problem.
```
Comment (ruscur) [inline content/blog/2023-08-03-parsing-ssh-config.markdown]: Is there a piece of advice you could leave people with in regards to writing their SSH config?

**Review by ruscur — APPROVED** [review]
is good

File: content/blog/2023-08-03-parsing-ssh-config.markdown (hunk 2 above)
Comment (BenjaminGrayNp1) [reply content/blog/2023-08-03-parsing-ssh-config.markdown:158]: Is this better? I don't know if it's necessary to point out the exact file and how it gets included

File: content/blog/2023-08-03-parsing-ssh-config.markdown (hunk 3 above)
Comment (BenjaminGrayNp1) [reply content/blog/2023-08-03-parsing-ssh-config.markdown]: Wisdom imparted

File: content/blog/2023-08-03-parsing-ssh-config.markdown (hunk 2 above)
Comment (ruscur) [reply content/blog/2023-08-03-parsing-ssh-config.markdown:158]: I don't think it's necessary I just think things like that help connect concepts to what someone is running in reality

File: content/blog/2023-08-03-parsing-ssh-config.markdown [hunk 4]
```diff
@@ -0,0 +1,229 @@
+Title: Quirks of parsing SSH configs 
+Date: 2023-08-03 00:00:00
```
Comment (ajdlinux) [inline content/blog/2023-08-03-parsing-ssh-config.markdown]: sorry to inform you that it is now the 4th

File: content/blog/2023-08-03-parsing-ssh-config.markdown [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+needs, but have never looked deeper into the actual rules behind how SSH parses
+the config.
+
+1. For starters, the config is parsed line by line. Leading whitespace (i.e.,
+   indentation) is ignored. So, while indentation makes it look like you are
+   configuring properties for a particular host, this isn't quite correct.
+   Instead, the `Host` and `Match` lines are special statements that enable or
+   disable all subsequent lines until the next `Host` or `Match`.
+
+    There is no backtracking; previous conditions and lines are not re-evaluated
```
Comment (ajdlinux) [inline content/blog/2023-08-03-parsing-ssh-config.markdown:106]: why is this indented wrong. idk if it changes the output, why is it wrong

File: content/blog/2023-08-03-parsing-ssh-config.markdown [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    after learning more about the config later on.
+
+2. When a config line is seen, and is active thanks to the most recent `Host` or
+   `Match` succeeding, its value is selected if it is the first of that config
+   to be selected. So the earliest place a value is set takes priority; this may
+   be a little counterintuitive if you are used to having the latest value be
+   picked, such as enable/disable command line flags tend to work.
```
Comment (ajdlinux) [inline content/blog/2023-08-03-parsing-ssh-config.markdown]: s/such as/like/

File: content/blog/2023-08-03-parsing-ssh-config.markdown [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+4. The last behaviour of interest is the `Match final` rule. There are several
+   conditions a `Match` statement can have, and the `final` rule says make this
+   active on the final pass over the config.
+
+Wait, final pass? Multiple passes? Yes. If `final` is a condition on a `Match`,
+SSH will do another pass over the config, following all the rules above. Except
+this time all the configs we read on the first pass are still active (and can't
+be changed). But all the `Host` and `Matches` are re-evaluated, allowing other
+configs to potentially be set. I guess that means rule (1) ought to have a big
```
Comment (ajdlinux) [inline content/blog/2023-08-03-parsing-ssh-config.markdown]: mention explicitly that the configs that can be affected aren't just the rules that are contained within the `Match final` as you go on to elaborate in the example below

File: content/blog/2023-08-03-parsing-ssh-config.markdown (hunk 2 above)
Comment (ajdlinux) [reply content/blog/2023-08-03-parsing-ssh-config.markdown:158]: I agree, you should talk about the concrete example and where the file is

File: content/blog/2023-08-03-parsing-ssh-config.markdown [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+that a given host name resolves to.
+
+Enter `ssh -G`. The `-G` flag asks SSH to dump the complete resolved config,
+without actually opening the connection (it may execute arbitrary code while
+resolving the config however!). So to fix the extension once and for all, we
+could swap the manual parser to just invoking `ssh -G example`, and parsing the
+output as the final config. No `Host` or `Match` or `HostName` or `Match final`
+quirks to worry about.
+
+Sure enough, if we replace the config backend with this 'native' resolver, we
```
Comment (ajdlinux) [inline content/blog/2023-08-03-parsing-ssh-config.markdown:212]: are you gonna upstream this

**Review by ajdlinux — COMMENTED** [review]
editorial comments

File: content/blog/2023-08-03-parsing-ssh-config.markdown (hunk 5 above)
Comment (BenjaminGrayNp1) [reply content/blog/2023-08-03-parsing-ssh-config.markdown:106]: 4 spaces required to be part of the bullet point. The indentation of the first paragraph was automatically done by the hard wrapper

File: content/blog/2023-08-03-parsing-ssh-config.markdown (hunk 8 above)
Comment (BenjaminGrayNp1) [reply content/blog/2023-08-03-parsing-ssh-config.markdown:212]: I've made a PR. And I'm working on fixing the parser too

Comment (ajdlinux) [reply content/blog/2023-08-03-parsing-ssh-config.markdown:212]: cool, throw in a link

File: content/blog/2023-08-03-parsing-ssh-config.markdown (hunk 5 above)
Comment (ajdlinux) [reply content/blog/2023-08-03-parsing-ssh-config.markdown:106]: huh ok

**Comment by BenjaminGrayNp1** [comment]
OK there are some mistakes in the `Host` rules. Turns out that during the first pass only the alias is used; `HostName` is only used as the host during the final pass (if requested). However, `Match host=...` uses `HostName` as soon as it is available.

**Pull request merged**